### PR TITLE
Docs for ListLookup

### PR DESCRIPTION
**Pull request by ducky64**
Document the beast that is ListLookup. It's something that seems to make a lot more sense with an example than can be described in words or implementation.

Additionally gives background for where this came from (chisel2, probably without much critical evaluation) and where this could go (deprecation).

**Related issue**: #411 

**Type of change**: other enhancement

**Impact**: no functional change

**Development Phase**: implementation

**Release Notes**
Scaladoc for ListLookup / Lookup.

### CODE REVIEW

File: src/main/scala/chisel3/util/Lookup.scala [hunk 1]
```diff
@@ -13,6 +36,18 @@ object ListLookup {
...
+/** Muxes between cases based on whether an address matches any pattern for a case.
+  * Similar to MuxLookup, but uses BitPat for address comparison.
+  *
+  * @note This appears to be an odd, specialized operator that we haven't seen used much, and seems
+  *       to be a holdover from chisel2. This may be deprecated and removed, usage is not
+  *       recommended.
+  *
+  * @param addr address to select between cases
+  * @param default default value should the address not match any case
+  * @mapping list of cases, where each entry consists of a BitPat (compared against addr) and the
```
Comment (seldridge) [inline src/main/scala/chisel3/util/Lookup.scala]:
```suggestion
  * @param mapping list of cases, where each entry consists of a BitPat (compared against addr) and the
```

File: src/main/scala/chisel3/util/Lookup.scala [hunk 2]
```diff
@@ -13,6 +36,18 @@ object ListLookup {
   }
 }
 
+/** Muxes between cases based on whether an address matches any pattern for a case.
+  * Similar to MuxLookup, but uses BitPat for address comparison.
```
Comment (seldridge) [inline src/main/scala/chisel3/util/Lookup.scala]:
```suggestion
  * Similar to [[chisel3.util.MuxLookup MuxLookup]], but uses [[chisel3.util.BitPat BitPat]] for address comparison.
```

File: src/main/scala/chisel3/util/Lookup.scala [hunk 3]
```diff
@@ -4,6 +4,29 @@ package chisel3.util
...
+/** For each element in a list, muxes (looks up) between cases (one per list element) based on a
+  * common address.
+  *
+  * @note This appears to be an odd, specialized operator that we haven't seen used much, and seems
+  *       to be a holdover from chisel2. This may be deprecated and removed, usage is not
+  *       recommended.
+  *
+  * @param addr common select for cases, shared (same) across all list elements
+  * @param default default value for each list element, should the address not match any case
+  * @param mapping list of cases, where each entry consists of a BitPat (compared against addr) and
```
Comment (seldridge) [inline src/main/scala/chisel3/util/Lookup.scala]:
```suggestion
  * @param mapping list of cases, where each entry consists of a [[chisel3.util.BitPat BitPath]] (compared against addr) and
```

File: src/main/scala/chisel3/util/Lookup.scala [hunk 4]
```diff
@@ -4,6 +4,29 @@ package chisel3.util
...
+  *                element (will have the same index in the output).
+  *
+  * @example {{{
+  *   ListLookup(1.U,  // address for comparison
+  *       List(      00.U, 01.U, 02.U)   // default "row" if none of the following cases match
+  *       List(1.U, (10.U, 11.U, 12.U),  // this "row" hardware-selected based off address 1.U
+  *            2.U, (20.U, 21.U, 22.U)
+  *       )
+  *   )  // hardware-evaluates to List(10.U, 11.U, 12.U)
+  *   // Note: if given address 0.U, the above would hardware evaluate to List(00.U, 01.U, 02.U)
```
Comment (edwardcwang) [inline src/main/scala/chisel3/util/Lookup.scala]: Not sure the types in this example are fully correct. I tried the following and it appears to work.

(Fun fact, `00.U` appears to be illegal)

```scala
ListLookup(2.U,  // address for comparison
                              List(10.U, 11.U, 12.U),   // default "row" if none of the following cases match
         Array(BitPat(2.U) -> List(20.U, 21.U, 22.U),  // this "row" hardware-selected based off address 2.U
               BitPat(3.U) -> List(30.U, 31.U, 32.U)
         )
     )
// hardware-evaluates to List(20.U, 21.U, 22.U)
```

Comment (ducky64) [reply src/main/scala/chisel3/util/Lookup.scala]: Fixed to your example. I guess that's what I get for not testing my code...

File: src/main/scala/chisel3/util/Lookup.scala [hunk 5]
```diff
@@ -4,6 +4,28 @@ package chisel3.util
...
+  *                a list of elements (same length as default) that is the output value for that
+  *                element (will have the same index in the output).
+  *
+  * @example {{{
+  * ListLookup(2.U,  // address for comparison
+  *                          List(10.U, 11.U, 12.U),   // default "row" if none of the following cases match
+  *     Array(BitPat(2.U) -> List(20.U, 21.U, 22.U),  // this "row" hardware-selected based off address 2.U
+  *           BitPat(3.U) -> List(30.U, 31.U, 32.U))
+  * ) // hardware-evaluates to List(20.U, 21.U, 22.U)
+  * // Note: if given address 0.U, the above would hardware evaluate to List(00.U, 01.U, 02.U)
```
Comment (edwardcwang) [inline src/main/scala/chisel3/util/Lookup.scala]:
```suggestion
  * // Note: if given address 0.U, the above would hardware evaluate to List(10.U, 11.U, 12.U)
```

Comment (ducky64) [reply src/main/scala/chisel3/util/Lookup.scala]: derrrrrp

**Pull request merged**